src/autoscaler/metricsforwarder/server/custom_metrics_handlers.go

```diff
@@ -0,0 +1,184 @@
+		if !isValid {
+			mh.logger.Error("error-validating-authorizaion-header", err, lager.Data{"authheader": auth})
+			http.Error(w, "Authorization failed", http.StatusUnauthorized)
+			handlers.WriteJSONResponse(w, http.StatusUnauthorized, models.ErrorResponse{
+				Code:    "Authorization-Failure-Error",
+				Message: "Basic authorization credential does not match"})
+			return
+		}
+		// Valid credentials found in DB so update the cache
+		mh.credentialCache.Set(appID, credentials, mh.cacheTTL)
+	}
+
+	body, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		mh.logger.Error("error-reading-request-body", err, lager.Data{"body": r.Body})
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		handlers.WriteJSONResponse(w, http.StatusInternalServerError, models.ErrorResponse{
+			Code:    "Interal-Server-Error",
+			Message: "Error reading custom metrics request body"})
+		return
+	}
+	var metricsConsumer *models.MetricsConsumer
+	err = json.Unmarshal(body, &metricsConsumer)
+	if err != nil {
+		mh.logger.Error("error-unmarshaling-metrics", err, lager.Data{"body": r.Body})
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		handlers.WriteJSONResponse(w, http.StatusInternalServerError, models.ErrorResponse{
+			Code:    "Interal-Server-Error",
+			Message: "Error unmarshaling custom metrics request body"})
+		return
+	}
+	isValidRequest, err := mh.policyDB.ValidateCustomMetricTypes(appID, metricsConsumer)
+	if !isValidRequest && err != nil {
+		mh.logger.Error("failed-validating-metrictypes", err, lager.Data{"metrics": metricsConsumer})
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		handlers.WriteJSONResponse(w, http.StatusBadRequest, models.ErrorResponse{
+			Code:    "Bad-Request",
+			Message: "Error validating custom metrics type"})
+		return
+	}
+
+	metrics := mh.parseMetrics(metricsConsumer)
+
+	if len(metrics) <= 0 {
+		mh.logger.Debug("failed-parsing-custom-metrics-request-body", lager.Data{"metrics": metrics})
+		handlers.WriteJSONResponse(w, http.StatusBadRequest, models.ErrorResponse{
+			Code:    "Bad-Request",
+			Message: "Error parsing request body"})
+		return
+	}
+
+	mh.logger.Debug("custom-metrics-parsed-successfully", lager.Data{"metrics": metrics})
+	for _, metric := range metrics {
+		mh.metricForwarder.EmitMetric(metric)
+	}
+	w.WriteHeader(http.StatusOK)
+}
+
+func (mh *CustomMetricsHandler) validateCredentials(username string, usernameHash string, password string, passwordHash string) bool {
+	isUsernameAuthenticated := bcrypt.CompareHashAndPassword([]byte(usernameHash), []byte(username))
+	isPasswordAuthenticated := bcrypt.CompareHashAndPassword([]byte(passwordHash), []byte(password))
+	if isPasswordAuthenticated == nil && isUsernameAuthenticated == nil { // password matching successfull
+		return true
+	}
+	mh.logger.Debug("failed-to-authorize-credentials")
+	return false
+}
+
+func (mh *CustomMetricsHandler) parseMetrics(metricsConsumer *models.MetricsConsumer) []*models.CustomMetric {
```

Comment:
this is actually not "parse", you may use "getMetrics"

Reply:
Renamed to getMetrics